[FreeType] Use builtin WOFF2 support

donny-dont opened the pull request:
#### ec5dd8737c21fc12b7e662e0cb8a5ef0e76998a6
<pre>
[FreeType] Use builtin WOFF2 support
<a href="https://bugs.webkit.org/show_bug.cgi?id=213862">https://bugs.webkit.org/show_bug.cgi?id=213862</a>

Reviewed by Michael Catanzaro.

FreeType v2.10.2 added native support for WOFF2. Add a CMake check to
verify that the support was compiled in. Enable this support for the
PlayStation port and make sure to load the brotli module if needed.

The only function used from `WOFFFileFormat.h` is
`convertWOFFToSfntIfNecessary` so only expose that. Clean up the
implementation of `WOFFFileFormat.cpp` to reflect that.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/platform/graphics/WOFFFileFormat.cpp:
(WebCore::convertWOFFToSfntIfNecessary):
* Source/WebCore/platform/graphics/WOFFFileFormat.h:
* Source/WebCore/platform/graphics/freetype/FontCustomPlatformDataFreeType.cpp:
(WebCore::FontCustomPlatformData::supportsFormat):
* Source/WebKit/WebProcess/EntryPoint/playstation/WebProcessMain.cpp:
(main):
* Source/cmake/OptionsPlayStation.cmake:
* Source/cmake/WOFF2Checks.cmake: Added.

Canonical link: <a href="https://commits.webkit.org/264815@main">https://commits.webkit.org/264815@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/25dcc746d6f9b25e245f862c3724486c50647836

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8689 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8979 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9194 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10346 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/8689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8698 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10967 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8950 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11564 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8835 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9853 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10503 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7153 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7948 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/15464 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/7468 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/8269 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8096 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11438 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/8335 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8576 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/6995 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/8862 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7841 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/2129 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2109 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12053 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/9091 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8321 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2249 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -472,6 +472,17 @@ if (USE_OPENJPEG)
 
+include(WOFF2Checks)
+if (FREETYPE_WOFF2_SUPPORT_IS_AVAILABLE)
+    SET_AND_EXPOSE_TO_BUILD(HAVE_WOFF_SUPPORT ON)
```
I think this should be `HAVE_FREETYPE_WOFF2_SUPPORT`, because at first I understood the flag as “we can have WOFF2 support” but it actually means “we can have WOFF2 support *from Freetype*”.

Other than that, I like it that the patch is simple and mostly about determining whether Freetype has the needed support. That would allow us to keep the fallback to `woff2dec` while supported distributions don't have the needed Freetype version.

I am making a local build with this applied to test it a bit. Thanks for the patch!

@clopez, @carlosgcampos: WDYT about this?

donny-dont replied:
I used that since CoreText has native support for WOFF(2). Open to different names but the idea was it was also applicable to Cocoa platforms.

mcatanzaro commented on Source/WebCore/platform/graphics/WOFFFileFormat.cpp:
```diff
@@ -25,18 +25,22 @@
 #include "SharedBuffer.h"
-#include <wtf/ByteOrder.h>
 
+#if !HAVE(WOFF_SUPPORT)
+#include <wtf/ByteOrder.h>
+#include <zlib.h>
```
Please avoid using conditionals to guard headers unless it's required. Otherwise you're just increasing the odds of non-unified build failures or build failures in odd conditions. E.g. what if another file in the same unified source bundle requires ByteOrder.h but forgets to include it?

donny-dont replied:
I agree with you just a note though the implementation for everything is gated on `!HAVE(WOFF_SUPPORT)` so in this case its perfectly fine.

mcatanzaro replied:
What I mean is: since the build won't fail if you remove the guard, remove it. Try to keep source files containing the same header files regardless of build options to the greatest extent possible. Minimizing includes is valuable but not at the cost of introducing more variability in what gets included in different configs.

donny-dont committed:
[FreeType] Use builtin WOFF2 support
https://bugs.webkit.org/show_bug.cgi?id=213862

Reviewed by Michael Catanzaro.

FreeType v2.10.2 added native support for WOFF2. Add a CMake check to
verify that the support was compiled in. Enable this support for the
PlayStation port and make sure to load the brotli module if needed.

The only function used from `WOFFFileFormat.h` is
`convertWOFFToSfntIfNecessary` so only expose that. Clean up the
implementation of `WOFFFileFormat.cpp` to reflect that.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/platform/graphics/WOFFFileFormat.cpp:
(WebCore::convertWOFFToSfntIfNecessary):
* Source/WebCore/platform/graphics/WOFFFileFormat.h:
* Source/WebCore/platform/graphics/freetype/FontCustomPlatformDataFreeType.cpp:
(WebCore::FontCustomPlatformData::supportsFormat):
* Source/WebKit/WebProcess/EntryPoint/playstation/WebProcessMain.cpp:
(main):
* Source/cmake/OptionsPlayStation.cmake:
* Source/cmake/WOFF2Checks.cmake: Added.

Canonical link: https://commits.webkit.org/264815@main
